# Project Capstone: COVID data analysis
## Data Cleaning

In this document, data cleaning tasks of the project are carried out.

The following datasets are used for the subsequent analysis:
- [Covid dataset from 'Over World in Data']( https://ourworldindata.org/coronavirus)
- [Covid measure dataset from 'The Humanitarian Data Exchange'](https://data.humdata.org/dataset/acaps-covid19-government-measures-dataset)

In case of interest, [here](https://www.marktechpost.com/2020/04/12/list-of-covid-19-resources-for-machine-learning-and-data-science-research/) you can find further COVID-19 datasets

In [50]:
# Loading required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [51]:
# Read csv with covid-19 cases and deaths
df_cases = pd.read_csv('data/owid-covid-data.csv')

# Read csv with covid-19 measures
df_measures = pd.read_excel('data/acaps_covid19_government_measures_datasets.xlsx', sheet_name = 1)

### OWID-COVID-DATA

In [52]:
# Let's start analyzing the df_cases
df_cases.sample(10)

iso_code      continent             location        date  total_cases  \
6908       FIN         Europe              Finland  2020-03-13          155   
20051      TTO  North America  Trinidad and Tobago  2020-03-29           74   
505        AGO         Africa               Angola  2020-04-22           24   
3491       KHM           Asia             Cambodia  2020-04-27          122   
18569      ESP         Europe                Spain  2020-03-01           82   
12601      MYS           Asia             Malaysia  2020-05-28         7619   
18350      KOR           Asia          South Korea  2020-03-01         3526   
14595      NIC  North America            Nicaragua  2020-04-29           13   
9192       ISL         Europe              Iceland  2020-06-04         1806   
11886      LTU         Europe            Lithuania  2020-02-02            0   

       new_cases  total_deaths  new_deaths  total_cases_per_million  \
6908          96             0           0                   27.975   
20051          8             2           0                   52.876   
505            0             2           0                    0.730   
3491           0             0           0                    7.297   
18569         28             0           0                    1.754   
12601         15           115           0                  235.401   
18350        595            17           1                   68.774   
14595          0             3           0                    1.962   
9192           0            10           0                 5292.308   
11886          0             0           0                    0.000   

       new_cases_per_million  ...  aged_65_older  aged_70_older  \
6908                  17.326  ...         21.228         13.264   
20051                  5.716  ...         10.014          5.819   
505                    0.000  ...          2.405          1.362   
3491                   0.000  ...          4.412          2.385   
18569                  0.599  ...         19.436         13.799   
12601                  0.463  ...          6.293          3.407   
18350                 11.605  ...         13.914          8.622   
14595                  0.000  ...          5.445          3.519   
9192                   0.000  ...         14.431          9.207   
11886                  0.000  ...         19.002         13.778   

       gdp_per_capita  extreme_poverty  cvd_death_rate  diabetes_prevalence  \
6908        40585.721              NaN         153.507                 5.76   
20051       28763.071              NaN         228.467                10.97   
505          5819.495              NaN         276.045                 3.94   
3491         3645.070              NaN         270.892                 4.00   
18569       34272.360              1.0          99.403                 7.17   
12601       26808.164              0.1         260.942                16.74   
18350       35938.374              0.2          85.998                 6.80   
14595        5321.444              3.2         137.016                11.47   
9192        46482.958              0.2         117.992                 5.31   
11886       29524.265              0.7         342.989                 3.67   

       female_smokers  male_smokers handwashing_facilities  \
6908             18.3          22.6                    NaN   
20051             NaN           NaN                 89.443   
505               NaN           NaN                 26.664   
3491              2.0          33.7                 66.229   
18569            27.4          31.4                    NaN   
12601             1.0          42.4                    NaN   
18350             6.2          40.9                    NaN   
14595             NaN           NaN                    NaN   
9192             14.3          15.2                    NaN   
11886            21.3          38.0                    NaN   

       hospital_beds_per_thousand  
6908                       

Every row represents a country and a date with the corresponding cases (new and cumulated), deaths (new and cumulated) and further facts.

In [53]:
df_cases.tests_units.sample(25)

384                  NaN
11905                NaN
12845                NaN
13902     samples tested
2062                 NaN
10077    tests performed
20629                NaN
17501      people tested
21140    tests performed
17060                NaN
12546      people tested
9333      samples tested
9522                 NaN
10376                NaN
8790                 NaN
20281    tests performed
21879                NaN
1901                 NaN
12819                NaN
4683                 NaN
5398                 NaN
14411                NaN
13398                NaN
18612    tests performed
13486                NaN
Name: tests_units, dtype: object

In [54]:
# Let's see columns, dtypes and nans
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22032 entries, 0 to 22031
Data columns (total 33 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   iso_code                         21968 non-null  object 
 1   continent                        21810 non-null  object 
 2   location                         22032 non-null  object 
 3   date                             22032 non-null  object 
 4   total_cases                      22032 non-null  int64  
 5   new_cases                        22032 non-null  int64  
 6   total_deaths                     22032 non-null  int64  
 7   new_deaths                       22032 non-null  int64  
 8   total_cases_per_million          21641 non-null  float64
 9   new_cases_per_million            21641 non-null  float64
 10  total_deaths_per_million         21641 non-null  float64
 11  new_deaths_per_million           21641 non-null  float64
 12  total_tests       

There are 33 different columns. We can find under the [readme cases dataframe](https://github.com/owid/covid-19-data/blob/master/public/data/owid-covid-data-codebook.md) all relevant information about every column . To sum up, it is possible to group the columns in three groups:
- Country and date information. (column 0 to 3)
- Deaths and cases, absolute and relative to the country population, as well as cumulative over time. (column 4 to 18)
- Country relevant information: population, population density, population age distribution, population disease information, gdp, etc.

Regarding data types, there are 5 string, 24 float and 5 int columns. We need just to parse the date, and have a closer look to the tests_units column which is also an object type.

Finally, looking at the nans and completeness, data looks quite clean. Just some issues to remark:
- From the 22032 rows, just columns location, date, total_cases, new_cases, total_deaths and new_deaths are complete. 
- Then, there is a set of columns with relative facts which have a few missing values. We should have a concrete look at these columns.
- When talking about the tests, we have just 25% of the values filled.
- Finally, columns with socio-economic information of the country have also some missing values.

In [55]:
# Parse date
df_cases.date = pd.to_datetime(df_cases.date, format = '%Y-%m-%d')

In [56]:
# Have a closer look at the tests_units column
df_cases.groupby(['tests_units'])['location'].count()

tests_units
people tested      1968
samples tested     1158
tests performed    3026
units unclear       935
Name: location, dtype: int64

Just a few rows are filled with information (30%). At the same time, the information provided by this column is not so comprehensible, even after reading the readme file. Consequently, this column is dropped.

In [57]:
# Drop tests_units column
df_cases.drop('tests_units', axis = 1, inplace = True)

In [58]:
df_cases.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths_per_million',
       'new_deaths_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'stringency_index', 'population', 'population_density', 'median_age',
       'aged_65_older', 'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cvd_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand'],
      dtype='object')

In [59]:
# Have a look at the socio-economic features. 
# Is the information just missing for some of the dates or just some of the countrys
# Starting with the columns information
df_cases[df_cases.population.isna()].groupby('location').count()

iso_code  continent  date  total_cases  new_cases  \
location                                                           
International         0          0    64           64         64   

               total_deaths  new_deaths  total_cases_per_million  \
location                                                           
International            64          64                        0   

               new_cases_per_million  total_deaths_per_million  ...  \
location                                                        ...   
International                      0                         0  ...   

               aged_65_older  aged_70_older  gdp_per_capita  extreme_poverty  \
location                                                                       
International              0              0               0                0   

               cvd_death_rate  diabetes_prevalence  female_smokers  \
location                                                             
International               0                    0               0   

               male_smokers  handwashing_facilities  \
location                                              
International             0                       0   

               hospital_beds_per_thousand  
location                                   
International                           0  

[1 rows x 31 columns]

Rows without information for the population are rows, whose country is international, which is not a country. Thus, all these rows are dropped, since they do not provide any relevant information at a country level analysis.

In [60]:
# Drop rows for international
df_cases = df_cases[df_cases.location != 'International']
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21968 entries, 0 to 21967
Data columns (total 32 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         21968 non-null  object        
 1   continent                        21810 non-null  object        
 2   location                         21968 non-null  object        
 3   date                             21968 non-null  datetime64[ns]
 4   total_cases                      21968 non-null  int64         
 5   new_cases                        21968 non-null  int64         
 6   total_deaths                     21968 non-null  int64         
 7   new_deaths                       21968 non-null  int64         
 8   total_cases_per_million          21641 non-null  float64       
 9   new_cases_per_million            21641 non-null  float64       
 10  total_deaths_per_million         21641 non-null  float64  

In [61]:
# Continuing with population density column
df_cases[df_cases.population_density.isna()].groupby('location').count()

iso_code  continent  date  total_cases  \
location                                                                  
Anguilla                               71         71    71           71   
Bonaire Sint Eustatius and Saba        65         65    65           65   
Falkland Islands                       63         63    63           63   
Guernsey                               78         78    78           78   
Jersey                                 78         78    78           78   
Montserrat                             76         76    76           76   
South Sudan                            61         61    61           61   
Syria                                  75         75    75           75   
Taiwan                                157        157   157          157   
Vatican                                83         83    83           83   
Western Sahara                         41         41    41           41   

                                 new_cases  total_deaths  new_deaths  \
location                                                               
Anguilla                                71            71          71   
Bonaire Sint Eustatius and Saba         65            65          65   
Falkland Islands                        63            63          63   
Guernsey                                78            78          78   
Jersey                                  78            78          78   
Montserrat                              76            76          76   
South Sudan                             61            61          61   
Syria                                   75            75          75   
Taiwan                                 157           157         157   
Vatican                                 83            83          83   
Western Sahara                          41            41          41   

                                 total_cases_per_million  \
location                                                   
Anguilla                                              71   
Bonaire Sint Eustatius and Saba                       65   
Falkland Islands                                      63   
Guernsey                                              78   
Jersey                                                78   
Montserrat                                            76   
South Sudan                                           61   
Syria                                                 75   
Taiwan                                               156   
Vatican                                               83   
Western Sahara                                        41   

                                 new_cases_per_million  \
location                                                 
Anguilla                                            71   
Bonaire Sint Eustatius and Saba                     65   
Falkland Islands                                    63   
Guernsey                                            78   
Jersey                                              78   
Montserrat                                          76   
South Sudan                                         61   
Syria                                               75   
Taiwan                                             156   
Vatican                                             83   
Western Sahara                                      41   

                                 total_deaths_per_million  ...  aged_65_older  \
location                                                   ...                  
Anguilla                                               71  ...              0   
Bonaire Sint Eustatius and Saba                        65  ...              0   
Falkland Islands                                       63  ...              0   
Guernsey                                               78  ...              0   
Jersey                                                 78  ...              0   
Montserrat       

This countries are really small ones, which could be considered as outlier countries due to its size and population.

In [72]:
# Thus, we drop this rows
country_list_to_drop = list(df_cases[df_cases.population_density.isna()].groupby('location').count().index)
df_cases = df_cases[~df_cases.location.isin(country_list_to_drop)]
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21120 entries, 0 to 21967
Data columns (total 32 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         21120 non-null  object        
 1   continent                        20962 non-null  object        
 2   location                         21120 non-null  object        
 3   date                             21120 non-null  datetime64[ns]
 4   total_cases                      21120 non-null  int64         
 5   new_cases                        21120 non-null  int64         
 6   total_deaths                     21120 non-null  int64         
 7   new_deaths                       21120 non-null  int64         
 8   total_cases_per_million          20794 non-null  float64       
 9   new_cases_per_million            20794 non-null  float64       
 10  total_deaths_per_million         20794 non-null  float64  

In [74]:
# Let's continue with the population age
df_cases[df_cases.median_age.isna()].groupby('location').count()

iso_code  continent  date  total_cases  new_cases  \
location                                                                       
Andorra                          84         84    84           84         84   
Bermuda                          78         78    78           78         78   
British Virgin Islands           71         71    71           71         71   
Cayman Islands                   78         78    78           78         78   
Dominica                         75         75    75           75         75   
Faeroe Islands                   78         78    78           78         78   
Gibraltar                        78         78    78           78         78   
Greenland                        78         78    78           78         78   
Isle of Man                      77         77    77           77         77   
Kosovo                           82         82    82           82         82   
Liechtenstein                    87         87    87           87         87   
Monaco                          146        146   146          146        146   
Northern Mariana Islands         67         67    67           67         67   
Saint Kitts and Nevis            72         72    72           72         72   
San Marino                      157        157   157          157        157   
Sint Maarten (Dutch part)        72         72    72           72         72   
Turks and Caicos Islands         73         73    73           73         73   

                           total_deaths  new_deaths  total_cases_per_million  \
location                                                                       
Andorra                              84          84                       84   
Bermuda                              78          78                       78   
British Virgin Islands               71          71                       71   
Cayman Islands                       78          78                       78   
Dominica                             75          75                       75   
Faeroe Islands                       78          78                       78   
Gibraltar                            78          78                       78   
Greenland                            78          78                       78   
Isle of Man                          77          77                       77   
Kosovo                               82          82                       82   
Liechtenstein                        87          87                       87   
Monaco                              146         146                      146   
Northern Mariana Islands             67          67                       67   
Saint Kitts and Nevis                72          72                       72   
San Marino                          157         157                      157   
Sint Maarten (Dutch part)            72          72                       72   
Turks and Caicos Islands             73          73                       73   

                           new_cases_per_million  total_deaths_per_million  \
location                                                                     
Andorra                                       84                        84   
Bermuda                                       78                        78   
British Virgin Islands                        71                        71   
Cayman Islands                                78                        78   
Dominica                                      75                        75   
Faeroe Islands                                78                        78   
Gibraltar                                     78                        78   
Greenland                                     78                        78   
Isle of Man                                   77                        77   
Kosovo                                        82                        82   
Liechtenstein                                 87 

We can follow a similar approach as previously, since again these countries are really small or just part of a bigger one (like Gibraltar or Isle of Man)

In [76]:
# Dropping these rows
country_list_to_drop = list(df_cases[df_cases.median_age.isna()].groupby('location').count().index)
df_cases = df_cases[~df_cases.location.isin(country_list_to_drop)]
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19667 entries, 0 to 21967
Data columns (total 32 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         19667 non-null  object        
 1   continent                        19509 non-null  object        
 2   location                         19667 non-null  object        
 3   date                             19667 non-null  datetime64[ns]
 4   total_cases                      19667 non-null  int64         
 5   new_cases                        19667 non-null  int64         
 6   total_deaths                     19667 non-null  int64         
 7   new_deaths                       19667 non-null  int64         
 8   total_cases_per_million          19341 non-null  float64       
 9   new_cases_per_million            19341 non-null  float64       
 10  total_deaths_per_million         19341 non-null  float64  

In [78]:
# Have a look at the aged_70_older
df_cases[df_cases.aged_70_older.isna()].groupby('location').count()

iso_code  continent  date  total_cases  new_cases  total_deaths  \
location                                                                    
Serbia         101        101   101          101        101           101   

          new_deaths  total_cases_per_million  new_cases_per_million  \
location                                                               
Serbia           101                       88                     88   

          total_deaths_per_million  ...  aged_65_older  aged_70_older  \
location                            ...                                 
Serbia                          88  ...            101              0   

          gdp_per_capita  extreme_poverty  cvd_death_rate  \
location                                                    
Serbia               101                0             101   

          diabetes_prevalence  female_smokers  male_smokers  \
location                                                      
Serbia                    101             101           101   

          handwashing_facilities  hospital_beds_per_thousand  
location                                                      
Serbia                       101                         101  

[1 rows x 31 columns]

In [89]:
# Having a look at neighbourhood countries, we can infered a value...
df_cases[df_cases.iso_code.isin(['HUN', 'BIH', 'ROU'])].groupby('location')['aged_70_older'].mean()

location
Bosnia and Herzegovina    10.711
Hungary                   11.976
Romania                   11.690
Name: aged_70_older, dtype: float64

In [117]:
mean_neigh = df_cases[df_cases.iso_code.isin(['HUN', 'BIH', 'ROU'])].groupby('location')['aged_70_older'].mean().mean()
index_list = df_cases[df_cases.location == 'Serbia'].index
df_cases.loc[index_list, 'aged_70_older'] = mean_neigh

In [123]:
# GDP colum
df_cases[df_cases.gdp_per_capita.isna()].groupby('location').count()                   

iso_code  continent  date  total_cases  \
location                                                               
Cuba                                83         83    83           83   
Curacao                             74         74    74           74   
French Polynesia                    79         79    79           79   
Guam                                79         79    79           79   
New Caledonia                       77         77    77           77   
Somalia                             81         81    81           81   
United States Virgin Islands        73         73    73           73   

                              new_cases  total_deaths  new_deaths  \
location                                                            
Cuba                                 83            83          83   
Curacao                              74            74          74   
French Polynesia                     79            79          79   
Guam                                 79            79          79   
New Caledonia                        77            77          77   
Somalia                              81            81          81   
United States Virgin Islands         73            73          73   

                              total_cases_per_million  new_cases_per_million  \
location                                                                       
Cuba                                               83                     83   
Curacao                                            74                     74   
French Polynesia                                   79                     79   
Guam                                               79                     79   
New Caledonia                                      77                     77   
Somalia                                            81                     81   
United States Virgin Islands                       73                     73   

                              total_deaths_per_million  ...  aged_65_older  \
location                                                ...                  
Cuba                                                83  ...             83   
Curacao                                             74  ...             74   
French Polynesia                                    79  ...             79   
Guam                                                79  ...             79   
New Caledonia                                       77  ...             77   
Somalia                                             81  ...             81   
United States Virgin Islands                        73  ...             73   

                              aged_70_older  gdp_per_capita  extreme_poverty  \
location                                                                       
Cuba                                     83               0                0   
Curacao                                  74               0                0   
French Polynesia                         79               0                0   
Guam                                     79               0                0   
New Caledonia                            77               0                0   
Somalia                                  81               0                0   
United States Virgin Islands             73               0                0   

                              cvd_death_rate  diabetes_prevalence  \
location                                                            
Cuba                                      83                   83   
Curacao                                    0                   74   
French Polynesia                           0                   79   
Guam                                      79                   79   
New Caledonia                              0                   77   
Somalia                                   81                   81   
United States Virgin Islands              73                  

Having a look in Google and in the World Bank Database, we can fulfill these missing values.

In [125]:
# Cuba
index_cuba = df_cases[df_cases.location == 'Cuba'].index
df_cases.loc[index_cuba, 'gdp_per_capita'] = 8821.82

# Curacao
index_curacao = df_cases[df_cases.location == 'Curacao'].index
df_cases.loc[index_curacao, 'gdp_per_capita'] = 19567.89

# French Polynesia
index_french = df_cases[df_cases.location == 'French Polynesia'].index
df_cases.loc[index_french, 'gdp_per_capita'] = 14323.82

# Guam
index_guam = df_cases[df_cases.location == 'Guam'].index
df_cases.loc[index_guam, 'gdp_per_capita'] = 35712.56

# New Caledonia
index_caledonia = df_cases[df_cases.location == 'New Caledonia'].index
df_cases.loc[index_caledonia, 'gdp_per_capita'] = 12579.60

# Somalia
index_somalia = df_cases[df_cases.location == 'Somalia'].index
df_cases.loc[index_somalia, 'gdp_per_capita'] = 314.54

# United States Virgin Islands
index_virgin = df_cases[df_cases.location == 'United States Virgin Islands'].index
df_cases.loc[index_virgin, 'gdp_per_capita'] = 35938.02

In [129]:
# cvd_death_rate column
df_cases[df_cases.cvd_death_rate.isna()].groupby('location').count() 

iso_code  continent  date  total_cases  new_cases  \
location                                                              
Aruba                   75         75    75           75         75   
Curacao                 74         74    74           74         74   
French Polynesia        79         79    79           79         79   
Hong Kong              123        123   123          123        123   
New Caledonia           77         77    77           77         77   

                  total_deaths  new_deaths  total_cases_per_million  \
location                                                              
Aruba                       75          75                       75   
Curacao                     74          74                       74   
French Polynesia            79          79                       79   
Hong Kong                  123         123                        0   
New Caledonia               77          77                       77   

                  new_cases_per_million  total_deaths_per_million  ...  \
location                                                           ...   
Aruba                                75                        75  ...   
Curacao                              74                        74  ...   
French Polynesia                     79                        79  ...   
Hong Kong                             0                         0  ...   
New Caledonia                        77                        77  ...   

                  aged_65_older  aged_70_older  gdp_per_capita  \
location                                                         
Aruba                        75             75              75   
Curacao                      74             74              74   
French Polynesia             79             79              79   
Hong Kong                   123            123             123   
New Caledonia                77             77              77   

                  extreme_poverty  cvd_death_rate  diabetes_prevalence  \
location                                                                 
Aruba                           0               0                   75   
Curacao                         0               0                   74   
French Polynesia                0               0                   79   
Hong Kong                       0               0                  123   
New Caledonia                   0               0                   77   

                  female_smokers  male_smokers  handwashing_facilities  \
location                                                                 
Aruba                          0             0                       0   
Curacao                        0             0                       0   
French Polynesia               0             0                       0   
Hong Kong                      0             0                       0   
New Caledonia                  0             0                       0   

                  hospital_beds_per_thousand  
location                                      
Aruba                                      0  
Curacao                                    0  
French Polynesia                           0  
Hong Kong                                  0  
New Caledonia                              0  

[5 rows x 31 columns]

All countries except Hong Kong are really small and can be considered as outliers. These rows are drop. For Hong Kong, the value can be inferred from a comparable country like Singapur.

In [135]:
# Dropping unrelevant countries
country_list_to_drop = ['Aruba', 'Curacao', 'French Polynesia', 'New Caledonia']
df_cases = df_cases[~df_cases.location.isin(country_list_to_drop)]

# Infered value from Singapur for Hong Kong
cvd_singapore = df_cases[df_cases.location == 'Singapore'].cvd_death_rate.mean()
index_singapore = df_cases[df_cases.location == 'Hong Kong'].index
df_cases.loc[index_singapore, 'cvd_death_rate'] = cvd_singapore

C:\Users\q426259\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [138]:
# Let's continue with smokers
df_cases[df_cases.female_smokers.isna()].groupby('location').count() 

iso_code  continent  date  total_cases  \
location                                                                   
Afghanistan                            148        148   148          148   
Angola                                  76         76    76           76   
Antigua and Barbuda                     78         78    78           78   
Belize                                  74         74    74           74   
Bhutan                                  84         84    84           84   
Bolivia                                 86         86    86           86   
Burundi                                 66         66    66           66   
Cameroon                                85         85    85           85   
Central African Republic                82         82    82           82   
Chad                                    78         78    78           78   
Cote d'Ivoire                           84         84    84           84   
Democratic Republic of Congo            85         85    85           85   
Equatorial Guinea                       83         83    83           83   
Gabon                                   84         84    84           84   
Grenada                                 75         75    75           75   
Guam                                    79         79    79           79   
Guatemala                               83         83    83           83   
Guinea                                  83         83    83           83   
Guinea-Bissau                           71         71    71           71   
Guyana                                  83         83    83           83   
Hong Kong                              123        123   123          123   
Iraq                                   156        156   156          156   
Jordan                                  85         85    85           85   
Libya                                   73         73    73           73   
Macedonia                              149        149   149          149   
Madagascar                              77         77    77           77   
Mauritania                              83         83    83           83   
Nicaragua                               79         79    79           79   
Palestine                               89         89    89           89   
Puerto Rico                             70         70    70           70   
Saint Lucia                             83         83    83           83   
Saint Vincent and the Grenadines        74         74    74           74   
Sao Tome and Principe                   58         58    58           58   
Somalia                                 81         81    81           81   
Sudan                                   84         84    84           84   
Tajikistan                              36         36    36           36   
Trinidad and Tobago                     85         85    85           85   
United States Virgin Islands            73         73    73           73   
Venezuela                               83         83    83           83   

                                  new_cases  total_deaths  new_deaths  \
location                                                                
Afghanistan                             148           148         148   
Angola                                   76            76          76   
Antigua and Barbuda                      78            78          78   
Belize                                   74            74          74   
Bhutan                                   84            84          84   
Bolivia                                  86            86          86   
Burundi                                  66            66          66   
Cameroon                                 85            85          85   
Central African Republic                 82            82          82   
Chad                                     78            78          78   
Cote d'Ivoire                            

In [140]:
df_cases[df_cases.male_smokers.isna()].groupby('location').count() 

iso_code  continent  date  total_cases  \
location                                                                   
Afghanistan                            148        148   148          148   
Angola                                  76         76    76           76   
Antigua and Barbuda                     78         78    78           78   
Belize                                  74         74    74           74   
Bhutan                                  84         84    84           84   
Bolivia                                 86         86    86           86   
Burundi                                 66         66    66           66   
Cameroon                                85         85    85           85   
Central African Republic                82         82    82           82   
Chad                                    78         78    78           78   
Cote d'Ivoire                           84         84    84           84   
Democratic Republic of Congo            85         85    85           85   
Equatorial Guinea                       83         83    83           83   
Gabon                                   84         84    84           84   
Grenada                                 75         75    75           75   
Guam                                    79         79    79           79   
Guatemala                               83         83    83           83   
Guinea                                  83         83    83           83   
Guinea-Bissau                           71         71    71           71   
Guyana                                  83         83    83           83   
Honduras                                84         84    84           84   
Hong Kong                              123        123   123          123   
Iraq                                   156        156   156          156   
Jordan                                  85         85    85           85   
Libya                                   73         73    73           73   
Macedonia                              149        149   149          149   
Madagascar                              77         77    77           77   
Mauritania                              83         83    83           83   
Nicaragua                               79         79    79           79   
Palestine                               89         89    89           89   
Peru                                    94         94    94           94   
Puerto Rico                             70         70    70           70   
Saint Lucia                             83         83    83           83   
Saint Vincent and the Grenadines        74         74    74           74   
Sao Tome and Principe                   58         58    58           58   
Somalia                                 81         81    81           81   
Sudan                                   84         84    84           84   
Tajikistan                              36         36    36           36   
Trinidad and Tobago                     85         85    85           85   
United States Virgin Islands            73         73    73           73   
Venezuela                               83         83    83           83   

                                  new_cases  total_deaths  new_deaths  \
location                                                                
Afghanistan                             148           148         148   
Angola                                   76            76          76   
Antigua and Barbuda                      78            78          78   
Belize                                   74            74          74   
Bhutan                                   84            84          84   
Bolivia                                  86            86          86   
Burundi                                  66            66          66   
Cameroon                                 85            85          85   
Central African Republic           

No clear pattern can be identified in this countries without information for the smokers rate. These values are not inferred to avoid corrupting the data. At the same time, we cannot drop all these columns, since we were loosing them a lot of information. Consequently, these nans are leave and in the future, depending on the applied algorithms, they will be handled.

A similar strategy will be followed for the rest of the columns with a great percentage of missing values.

In [141]:
# Continent column
df_cases[df_cases.continent.isna()]

iso_code continent location       date  total_cases  new_cases  \
21810  OWID_WRL       NaN    World 2019-12-31           27         27   
21811  OWID_WRL       NaN    World 2020-01-01           27          0   
21812  OWID_WRL       NaN    World 2020-01-02           27          0   
21813  OWID_WRL       NaN    World 2020-01-03           44         17   
21814  OWID_WRL       NaN    World 2020-01-04           44          0   
...         ...       ...      ...        ...          ...        ...   
21963  OWID_WRL       NaN    World 2020-06-01      6136294     107968   
21964  OWID_WRL       NaN    World 2020-06-02      6236775     100481   
21965  OWID_WRL       NaN    World 2020-06-03      6349294     112519   
21966  OWID_WRL       NaN    World 2020-06-04      6475978     126684   
21967  OWID_WRL       NaN    World 2020-06-05      6603329     127351   

       total_deaths  new_deaths  total_cases_per_million  \
21810             0           0                    0.003   
21811             0           0                    0.003   
21812             0           0                    0.003   
21813             0           0                    0.006   
21814             0           0                    0.006   
...             ...         ...                      ...   
21963        372667        3723                  787.229   
21964        376244        3577                  800.120   
21965        380803        4559                  814.555   
21966        386544        5741                  830.808   
21967        391732        5188                  847.146   

       new_cases_per_million  ...  aged_65_older  aged_70_older  \
21810                  0.003  ...          8.696          5.355   
21811                  0.000  ...          8.696          5.355   
21812                  0.000  ...          8.696          5.355   
21813                  0.002  ...          8.696          5.355   
21814                  0.000  ...          8.696          5.355   
...                      ...  ...            ...            ...   
21963                 13.851  ...          8.696          5.355   
21964                 12.891  ...          8.696          5.355   
21965                 14.435  ...          8.696          5.355   
21966                 16.252  ...          8.696          5.355   
21967                 16.338  ...          8.696          5.355   

       gdp_per_capita  extreme_poverty  cvd_death_rate  diabetes_prevalence  \
21810       15469.207             10.0          233.07                 8.51   
21811       15469.207             10.0          233.07                 8.51   
21812       15469.207             10.0          233.07                 8.51   
21813       15469.207             10.0          233.07                 8.51   
21814       15469.207             10.0          233.07                 8.51   
...               ...              ...             ...                  ...   
21963       15469.207             10.0          233.07                 8.51   
21964       15469.207             10.0          233.07                 8.51   
21965       15469.207             10.0          233.07                 8.51   
21966       15469.207             10.0          233.07                 8.51   
21967       15469.207             10.0          233.07                 8.51   

       female_smokers  male_smokers  handwashing_facilities  \
21810           6.434        34.635                   60.13   
21811           6.434        34.635                   60.13   
21812           6.434        34.635                   60.13   
21813           6.434        34.635                   60.13   
21814           6.434        34.635                   60.13   
...               ...           ...                     ...   
21963           6.434        34.635                   60.13   
21964           6.434        34.635                   60.13   
21965           6.434        34.635                   60.13   
21966           6.434    

Rows with missing information for the continent is the location "World", which cannot be ordered to a concrete continent. Thus, since this info could be valuable to analyze the global trend, no measure are taken for the continent column.

In [147]:
# Have a look at total_cases_per_million
df_cases[df_cases.total_cases_per_million.isna()].groupby('location').mean()

total_cases  new_cases  total_deaths  new_deaths  \
location                                                         
Bangladesh              0          0             0           0   
Belarus                 0          0             0           0   
Bolivia                 0          0             0           0   
Colombia                0          0             0           0   
Croatia                 0          0             0           0   
Estonia                 0          0             0           0   
Finland                 0          0             0           0   
Greece                  0          0             0           0   
Hong Kong               0          0             0           0   
Hungary                 0          0             0           0   
Ireland                 0          0             0           0   
Israel                  0          0             0           0   
Kazakhstan              0          0             0           0   
Kenya                   0          0             0           0   
Latvia                  0          0             0           0   
Lithuania               0          0             0           0   
Malaysia                0          0             0           0   
Mexico                  0          0             0           0   
Morocco                 0          0             0           0   
Nepal                   0          0             0           0   
New Zealand             0          0             0           0   
Panama                  0          0             0           0   
Paraguay                0          0             0           0   
Peru                    0          0             0           0   
Poland                  0          0             0           0   
Portugal                0          0             0           0   
Senegal                 0          0             0           0   
Serbia                  0          0             0           0   
South Africa            0          0             0           0   
Vietnam                 0          0             0           0   

              total_cases_per_million  new_cases_per_million  \
location                                                       
Bangladesh                        NaN                    NaN   
Belarus                           NaN                    NaN   
Bolivia                           NaN                    NaN   
Colombia                          NaN                    NaN   
Croatia                           NaN                    NaN   
Estonia                           NaN                    NaN   
Finland                           NaN                    NaN   
Greece                            NaN                    NaN   
Hong Kong                         NaN                    NaN   
Hungary                           NaN                    NaN   
Ireland                           NaN                    NaN   
Israel                            NaN                    NaN   
Kazakhstan                        NaN                    NaN   
Kenya                             NaN                    NaN   
Latvia                            NaN                    NaN   
Lithuania                         NaN                    NaN   
Malaysia                          NaN                    NaN   
Mexico                            NaN                    NaN   
Morocco                           NaN                    NaN   
Nepal                             NaN                    NaN   
New Zealand                       NaN                    NaN   
Panama                            NaN                    NaN   
Paraguay                          NaN                    NaN   
Peru                              NaN                    NaN   
Poland                            NaN                    NaN   
Portugal                          NaN                    NaN   
Senegal                           NaN                    NaN   
Serbia                            NaN        

All these rows are characterized by having 0 cases and deaths. Thus, it can be concluded that these rows correspond to days prior to the pandemy in the respective countries or without records for these countries. Since we have the amount of cases and deaths for these countries declared as 0, a value of 0 will be inferred for the relative features.

In [151]:
index_to_inferred_relative = list(df_cases[df_cases.total_cases_per_million.isna()].index)
columns_relative = ['total_cases_per_million', 'new_cases_per_million', 'total_deaths_per_million', 'new_deaths_per_million']
df_cases.loc[index_to_inferred_relative, columns_relative] = 0

C:\Users\q426259\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [152]:
# Have a look at the resulting df
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19362 entries, 0 to 21967
Data columns (total 32 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         19362 non-null  object        
 1   continent                        19204 non-null  object        
 2   location                         19362 non-null  object        
 3   date                             19362 non-null  datetime64[ns]
 4   total_cases                      19362 non-null  int64         
 5   new_cases                        19362 non-null  int64         
 6   total_deaths                     19362 non-null  int64         
 7   new_deaths                       19362 non-null  int64         
 8   total_cases_per_million          19362 non-null  float64       
 9   new_cases_per_million            19362 non-null  float64       
 10  total_deaths_per_million         19362 non-null  float64  

Now we can say that we have a cleaned dataset to work with. Most of the rows are completed, while some of them regarding tests, smoking behaviours, stringency_index, poverty, handwashing_facilities and hospital_beds_per_thousand have missing varlues which cannot be inferred without manipulating/corrupting the data significantly.
Consequently, we will work with these missing values and in case the applied methods do not allow them, a local solution just for this case will be applied.

In [155]:
# Save the clean dataset
df_cases.to_csv('data/owid-covid-data_clean.csv')

### acaps_covid19_government_measures_datasets

In [159]:
# Have a look at the dataset
df_measures.head()

ID      COUNTRY  ISO ADMIN_LEVEL_NAME  PCODE REGION  \
0   1  Afghanistan  AFG              NaN    NaN   Asia   
1   2  Afghanistan  AFG            Kabul    NaN   Asia   
2   3  Afghanistan  AFG              NaN    NaN   Asia   
3   4  Afghanistan  AFG              NaN    NaN   Asia   
4   5  Afghanistan  AFG              NaN    NaN   Asia   

                               LOG_TYPE  \
0  Introduction / extension of measures   
1  Introduction / extension of measures   
2  Introduction / extension of measures   
3  Introduction / extension of measures   
4  Introduction / extension of measures   

                                 CATEGORY  \
0                  Public health measures   
1                  Public health measures   
2                  Public health measures   
3  Governance and socio-economic measures   
4                       Social distancing   

                                             MEASURE TARGETED_POP_GROUP  \
0  Health screenings in airports and border cross...                 No   
1                  Isolation and quarantine policies                 No   
2                                Awareness campaigns                 No   
3  Emergency administrative structures activated ...                 No   
4                            Limit public gatherings                 No   

                    COMMENTS NON_COMPLIANCE DATE_IMPLEMENTED  \
0                        NaN            NaN       2020-02-12   
1                        NaN            NaN       2020-02-12   
2                        NaN            NaN       2020-02-12   
3                        NaN            NaN       2020-02-12   
4  Nevruz festival cancelled            NaN       2020-03-12   

               SOURCE SOURCE_TYPE  \
0  Ministry of Health  Government   
1  Ministry of Health  Government   
2  Ministry of Health  Government   
3  Ministry of Health  Government   
4                  AA       Media   

                                                LINK ENTRY_DATE  \
0  https://moph.gov.af/en/moph-held-emergency-mee... 2020-03-14   
1  https://moph.gov.af/en/moph-held-emergency-mee... 2020-03-14   
2  https://moph.gov.af/en/moph-held-emergency-mee... 2020-03-14   
3  https://moph.gov.af/en/moph-held-emergency-mee... 2020-03-14   
4  https://www.aa.com.tr/en/asia-pacific/coronavi... 2020-03-14   

  Alternative source  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN

In [160]:
# Dtypes, nans and no. of rows
df_measures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12538 entries, 0 to 12537
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   ID                  12538 non-null  int64         
 1   COUNTRY             12538 non-null  object        
 2   ISO                 12538 non-null  object        
 3   ADMIN_LEVEL_NAME    1323 non-null   object        
 4   PCODE               0 non-null      float64       
 5   REGION              12538 non-null  object        
 6   LOG_TYPE            12538 non-null  object        
 7   CATEGORY            12538 non-null  object        
 8   MEASURE             12538 non-null  object        
 9   TARGETED_POP_GROUP  12538 non-null  object        
 10  COMMENTS            12392 non-null  object        
 11  NON_COMPLIANCE      11276 non-null  object        
 12  DATE_IMPLEMENTED    12251 non-null  datetime64[ns]
 13  SOURCE              12518 non-null  object    

This dataset looks cleaner than the previous one. Most of the columns are complete except a few of them, which have just few missing values. Column PCODE is complete empty. Thus, we can drop it without loosing any information.

Regarding the meaning of every column and possible values, the [readme file](https://s3.eu-central-1.amazonaws.com/hdx-ckan-filestore-prod/resources/2d7b46c4-22fd-4480-b14e-d0a6899256c4/acaps-covid-19-government-measures-dataset-readme.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=180&X-Amz-Credential=AKIARZNKTAO7XQY7ED6N%2F20200606%2Feu-central-1%2Fs3%2Faws4_request&X-Amz-SignedHeaders=host&X-Amz-Date=20200606T065230Z&X-Amz-Signature=8c01c14ff4a1ea0b94c8c860fdd2b850647afa92b4f07320f9801fc5218805d3) provided with the dataset is quite useful.

With regard to data types, most of the columns are strings, while there are one int column, 2 datetime columns and a float column, which will be dropped.

Another relevant fact is to have a common key with the previous dataset to merge in case needed for a deeper analysis. The ISO country name and dates can be used for this purpose. Country name could be also used. However, there could be some issues in case of missspellings, for example.

In [162]:
# Drop pcode column
df_measures.drop('PCODE', axis = 1, inplace = True)

In [170]:
# Have a look at the admin column
df_measures.groupby('ADMIN_LEVEL_NAME')['ID'].count().to_frame()

ID
ADMIN_LEVEL_NAME                                      
                                                     1
 Al-Qatif                                            1
 Al-Qatif                                            1
 Batu 21 until Batu 24 in Sungai Lui in Selangor     1
 Federation of Bosnia and Herzegovina                1
...                                                 ..
parish of St. Catherine                              2
regions of Gabu, Oio and Bafata                      1
travellers                                           1
west bank                                            1
whole Central province, covering the National C...   3

[649 rows x 1 columns]

The information from this column is quite country specific and difficult to extract any value from it or extrapolate any trend. Thus, the columns will be dropped.

In [171]:
# Drop ADMIN_LEVEL_NAME
df_measures.drop('ADMIN_LEVEL_NAME', axis = 1, inplace = True)

In [173]:
# Having a look at the comments column
df_measures.COMMENTS.sample(25)

3109     Exemptions from the obligation to wear veils  ...
5462     Higher Committee for Health and National Safet...
5289     CBDT issues orders u/s 119 of IT Act,1961 to m...
6048     Kenyan gvmt has issued a travel advisory to al...
280      individual sports will be allowed from 6am to ...
1067     The Ministry of Health and Wellness has advise...
10434    The weekly flight from Brisbane (Australia) to...
6605     Building material stores have been placed unde...
8303     Several measures aimed at reducing the economi...
7232     The Rukun Tetangga neighbourhood watch is bein...
3070           Re-opening of public administration offices
1860     All food premises including restaurants, eater...
4854                                                   NaN
6550     Travellers from China and other COVID-19 affec...
2069     Civil servants or civil servants, as well as p...
1875       Guidelines for the celebration of Eid Aidifitri
5445     authorities banned travel between all Iraqi pr.

This column contains further comments for the measure in case it was neccesary. In case some NLP methods want to be applied to the dataset, this information could be interesting. NaNs can be interpreted as not necessary further explanations.

In [175]:
# Non compliance column: data entered with a dropdown
df_measures.NON_COMPLIANCE.sample(25)

987        Not available
4219               Fines
10257              Fines
5458     Up to detention
687       Not applicable
1869      Not applicable
2344      Not applicable
535       Not applicable
9517      Not applicable
8864      Not applicable
9565       Not available
4511      Not applicable
4461      Not applicable
3521      Not applicable
9188      Not applicable
1133      Not applicable
4455      Not applicable
6156                 NaN
10234      Not available
3169      Not applicable
4041      Not applicable
1098      Not applicable
1087      Not applicable
1706                 NaN
5513                 NaN
Name: NON_COMPLIANCE, dtype: object

Since the category not available is provided, this will be inferred for nans.

In [181]:
# Inferring not available for nans
index_nans = df_measures[df_measures.NON_COMPLIANCE.isna()].index
df_measures.loc[index_nans, 'NON_COMPLIANCE'] = 'Not available'

In [184]:
# Date implemented
df_measures[df_measures.DATE_IMPLEMENTED.isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 287 entries, 89 to 12440
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   ID                  287 non-null    int64         
 1   COUNTRY             287 non-null    object        
 2   ISO                 287 non-null    object        
 3   REGION              287 non-null    object        
 4   LOG_TYPE            287 non-null    object        
 5   CATEGORY            287 non-null    object        
 6   MEASURE             287 non-null    object        
 7   TARGETED_POP_GROUP  287 non-null    object        
 8   COMMENTS            269 non-null    object        
 9   NON_COMPLIANCE      287 non-null    object        
 10  DATE_IMPLEMENTED    0 non-null      datetime64[ns]
 11  SOURCE              280 non-null    object        
 12  SOURCE_TYPE         281 non-null    object        
 13  LINK                276 non-null    object     

For all these rows, where no date_implemented is provided, we have always a entry_date. Thus, it will be assumed that the entry_date = date_implemented.

In [186]:
# Infering date_implemented = entry_date if date_implemented not available
index_date_implemented = list(df_measures[df_measures.DATE_IMPLEMENTED.isna()].index)
df_measures.loc[index_date_implemented, 'DATE_IMPLEMENTED'] = df_measures.loc[index_date_implemented, 'ENTRY_DATE']

The rest of the columns (source, source_type, link and alternative source) say just from where the information was extracted and it is not relevant for the analysis. Thus, are dropped.

In [188]:
df_measures.drop(['SOURCE', 'SOURCE_TYPE', 'LINK', 'Alternative source'], axis = 1, inplace = True)

In [189]:
# Have a final look to the df
df_measures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12538 entries, 0 to 12537
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   ID                  12538 non-null  int64         
 1   COUNTRY             12538 non-null  object        
 2   ISO                 12538 non-null  object        
 3   REGION              12538 non-null  object        
 4   LOG_TYPE            12538 non-null  object        
 5   CATEGORY            12538 non-null  object        
 6   MEASURE             12538 non-null  object        
 7   TARGETED_POP_GROUP  12538 non-null  object        
 8   COMMENTS            12392 non-null  object        
 9   NON_COMPLIANCE      12538 non-null  object        
 10  DATE_IMPLEMENTED    12538 non-null  datetime64[ns]
 11  ENTRY_DATE          12538 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(9)
memory usage: 1.1+ MB


Now we have a clean dataset to work with. Let's save it for the future.

In [192]:
df_measures.to_csv('data/acaps_covid19_government_measures_datasets_clean.csv')